### PR TITLE
feat: Clear session for deleted users

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

- Email verification endpoint now clears the session if called with a deleted/unknown user
- Sign out now always clears cookies - even if there is no valid session attached

## Related issues

-   https://github.com/supertokens/supertokens-auth-react/issues/580

## Test Plan

Added tests

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py

### CODE REVIEW

File: supertokens_python/recipe/emailverification/recipe.py [hunk 1]
```diff
@@ -313,9 +314,7 @@ async def fetch_value(user_id: str, user_context: Dict[str, Any]) -> bool:
             if isinstance(email_info, EmailDoesNotExistError):
                 # we consider people without email addresses as validated
                 return True
-            raise Exception(
-                "Should never come here: UNKNOWN_USER_ID or invalid result from get_email_for_user"
-            )
+            raise_unauthorised_exception("UNKNOWN_USER_ID")  # todo: empty in node SDK
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:317]: - As per the node PR, this should throw a regular exception with the UNKNOWN_USER_ID message
- Why is there a TODO next to it? Is something wrong with the node SDK?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 2]
```diff
@@ -341,7 +340,14 @@ async def email_verify_post(
...
         if isinstance(response, VerifyEmailUsingTokenOkResult):
             if session is not None:
-                await session.fetch_and_set_claim(EmailVerificationClaim, user_context)
+                try:
+                    await session.fetch_and_set_claim(
+                        EmailVerificationClaim, user_context
+                    )
+                except Exception as e:
+                    # This should never happen since we have just set thestatus above
+                    if isinstance(e, UnknownUserIdError):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:349]: Change to check that the exception message to check if the message is `UNKNOWN_USER_ID`

Comment (rishabhpoddar) [reply supertokens_python/recipe/emailverification/recipe.py:349]: If it's not that, you need to rethrow `e`

File: supertokens_python/recipe/emailverification/recipe.py [hunk 3]
```diff
@@ -354,15 +360,18 @@ async def is_email_verified_get(
...
+
         is_verified = await session.get_claim_value(
             EmailVerificationClaim, user_context
         )
 
         if is_verified is None:
-            raise Exception(
-                "Should never come here: EmailVerificationClaim failed to set value"
-            )
+            raise_unauthorised_exception("Unknown User ID provided")
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:375]: this is wrong

Comment (rishabhpoddar) [reply supertokens_python/recipe/emailverification/recipe.py:375]: This should be in the generate_email_verify_token_post function

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/emailverification/recipe.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:349]: > If it's not that, you need to rethrow e

I did this. But I don't think this is happening with the Node SDK.

File: supertokens_python/recipe/emailverification/recipe.py [hunk 4]
```diff
@@ -313,9 +318,7 @@ async def fetch_value(user_id: str, user_context: Dict[str, Any]) -> bool:
             if isinstance(email_info, EmailDoesNotExistError):
                 # we consider people without email addresses as validated
                 return True
-            raise Exception(
-                "Should never come here: UNKNOWN_USER_ID or invalid result from get_email_for_user"
-            )
+            raise GeneralError("UNKNOWN_USER_ID")
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:321]: Just raise normal exception, not GeneralError

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**